Source/WebCore/html/CanvasBase.cpp

```diff
@@ -373,6 +374,203 @@ bool CanvasBase::shouldInjectNoiseBeforeReadback() const
+        tightestBoundingIndices = { indexOf(aboveLeftOffset), indexOf(belowRightOffset) };
+    }
+
+    diff = aboveRight - belowLeft;
+    if (areColorsInOrder(aboveRight, color, belowLeft) && diff < tightestBoundingDiff) {
+        tightestBoundingDiff = diff;
+        tightestBoundingIndices = { indexOf(aboveRightOffset), indexOf(belowLeftOffset) };
+    }
+
+    return tightestBoundingIndices;
+}
+
+static RGBAColor clampedColor(int tweak, const RGBAColor& color, const RGBAColor& boundingColor1, const RGBAColor& boundingColor2)
+{
+    constexpr auto minColorValue = static_cast<int32_t>(std::numeric_limits<uint8_t>::min());
+    constexpr auto maxColorValue = static_cast<int32_t>(std::numeric_limits<uint8_t>::max());
+
+    auto isNeighbor1LessThanNeighbor2 = boundingColor1 < boundingColor2;
+    const auto& smallerColor = isNeighbor1LessThanNeighbor2 ? boundingColor1 : boundingColor2;
+    const auto& largerColor = isNeighbor1LessThanNeighbor2 ? boundingColor2 : boundingColor1;
+    bool isBlack = !(color.red || color.green || color.blue);
+
+    uint8_t clampedRed { 0 };
+    uint8_t clampedGreen { 0 };
+    uint8_t clampedBlue { 0 };
+    uint8_t clampedAlpha { 0 };
+
+    if (!isBlack) {
+        clampedRed = std::max({ std::min({ color.red + tweak, static_cast<int>(largerColor.red), maxColorValue }), static_cast<int>(smallerColor.red), minColorValue });
+        clampedGreen = std::max({ std::min({ color.green + tweak, static_cast<int>(largerColor.green), maxColorValue }), static_cast<int>(smallerColor.green), minColorValue });
+        clampedBlue = std::max({ std::min({ color.blue + tweak, static_cast<int>(largerColor.blue), maxColorValue }), static_cast<int>(smallerColor.blue), minColorValue });
+    }
+    clampedAlpha = std::max({ std::min({ color.alpha + tweak, static_cast<int>(largerColor.alpha), maxColorValue }), static_cast<int>(smallerColor.alpha), minColorValue });
+
+    return {
+        clampedRed,
+        clampedGreen,
+        clampedBlue,
+        clampedAlpha
+    };
+}
+
+static void setNewColorInPixelBuffer(const Vector<size_t>& pixelIndices, auto tweakedColor, uint8_t* bytes, int size)
+{
+    for (auto index : pixelIndices) {
+        if (!isIndexInBounds(size, index) || !isIndexInBounds(size, index + 3))
+            continue;
+        bytes[index] = tweakedColor.red;
+        bytes[index + 1] = tweakedColor.green;
+        bytes[index + 2] = tweakedColor.blue;
+        bytes[index + 3] = tweakedColor.alpha;
+    }
+}
+
+static uint32_t computeOnePercentColorTweak(const RGBAColor& rgbaColor, uint64_t salt)
+{
+    uint32_t tweak { 0 };
+    uint64_t pixelHash = computeHash(salt, rgbaColor.red, rgbaColor.green, rgbaColor.blue, rgbaColor.alpha);
+    while (!tweak) {
+        pixelHash = computeHash(pixelHash);
+        // +/- 3 is roughly ~1% of the 255 max value.
+        tweak = static_cast<uint32_t>(((pixelHash * 6) / std::numeric_limits<uint32_t>::max()) - 3);
+    }
+    return tweak;
+}
+
```

Comment:
This noise algorithm now adds significant noise to already complex CanvasBase, doubling its size, and still it's a bit peripheral to the Canvas functionality.

Could we encapsulate the noise injection into a new file, such as CanvasNoiseInjection.h/cpp?